### ISSUE TITLE
Misc encoding problems with Non-ASCII files (ISO-8859-1)

### ISSUE THREAD

**Issue by danielmarschall** [viewvc/viewvc#165]
I had a lot of problems using code which had comments in my native language (encoded in ISO-8859-1),
because UTF-8 was always chosen as default encoding. I have changed the code a bit and added
more configuration switches to allow the administrator to have more influence in the encoding that is being used.

I have also found a spot where the config value detect_encoding was not queries correctly.
Also, another bug was that the encoding-search-algorithm was only available in the plain text files, not in the pygmentized code.

Here is a patch I have written. Can somebody please create a PullRequest?

```
diff -U 2 -r viewvc_original/conf/viewvc.conf.dist viewvc/conf/viewvc.conf.dist
--- viewvc_original/conf/viewvc.conf.dist	2018-11-01 11:06:14.824994030 +0100
+++ viewvc/conf/viewvc.conf.dist	2018-11-02 18:10:17.305992089 +0100
@@ -349,4 +349,15 @@
 [options]
 
+## If the algorithm has problems detecting the encoding of a file, try the system locale (if available)
+encoding_try_locale = 1
+
+## If the detection of the encoding failed, choose this encoding as last resort
+encoding_default = iso-8859-1
+
 ## root_as_url_component: Interpret the first path component in the URL
 ## after the script location as the root to use.  This is an

diff -U 2 -r viewvc_original/lib/viewvc.py viewvc/lib/viewvc.py
--- viewvc_original/lib/viewvc.py	2018-11-01 11:06:14.836993859 +0100
+++ viewvc/lib/viewvc.py	2018-11-01 23:27:40.601850575 +0100
@@ -39,4 +39,6 @@
 import urllib
 
+import locale
+
 # These modules come from our library (the stub has set up the path)
 from common import _item, _RCSDIFF_NO_CHANGES, _RCSDIFF_IS_BINARY, _RCSDIFF_ERROR, TemplateData
@@ -1697,10 +1718,10 @@
 
 
-def detect_encoding(text_block):
+def detect_encoding(text_block, cfg):
   """Return the encoding used by TEXT_BLOCK as detected by the chardet
   Python module.  (Currently, this is used only when syntax
   highlighting is not enabled/available; otherwise, Pygments does this
   work for us.)"""

   # Does the TEXT_BLOCK start with a BOM?
   for bom, encoding in [('\xef\xbb\xbf', 'utf-8'),
@@ -1714,21 +1735,30 @@
 
   # If no recognized BOM, see if chardet can help us.
-  try:
-    import chardet
+  if cfg.options.detect_encoding:
+    try:
+      import chardet
 
-    # If chardet can confidently claimed a match, we'll use its
-    # findings.  (And if that match is 'ascii' -- which is a subset of
-    # utf-8 -- we'll just call it 'utf-8' and score a zero transform.)
-    resp = chardet.detect(text_block)
-    if resp.get('confidence') == 1.0:
-      encoding = resp.get('encoding')
-      if encoding is "ascii":
-        encoding = "utf-8"
-      return encoding
-  except:
-    pass
+      # If chardet can confidently claimed a match, we'll use its
+      # findings.  (And if that match is 'ascii' -- which is a subset of
+      # utf-8 -- we'll just call it 'utf-8' and score a zero transform.)
+      resp = chardet.detect(text_block)
+      if resp.get('confidence') == 1.0:
+        encoding = resp.get('encoding')
+        if encoding is "ascii":
+          encoding = cfg.options.encoding_default
+        return encoding
+    except:
+      pass
+
+  # We try the system's default encoding
+  if cfg.options.encoding_try_locale:
+    tmp = locale.getdefaultlocale()[1]
+    if tmp != None:
+      return tmp
 
-  # By default ... we have no idea.
-  return None
+  return cfg.options.encoding_default
   
 def transcode_text(text, encoding=None):
@@ -1750,4 +1780,19 @@
     return []
 
+  # If allowed by configuration, try to detect the source encoding
+  # for this file.  We'll assemble a block of data from the file
+  # contents to do so... 1024 bytes should be enough.
+  if not encoding:
+    if cfg.options.detect_encoding:
+      block_size = 0
+      text_block = ''
+      for i in range(len(file_lines)):
+        text_block = text_block + file_lines[i]
+        if len(text_block) >= 1024:
+          break
+      encoding = detect_encoding(text_block, cfg)
+    else:
+      encoding = cfg.options.encoding_default
+
   # Determine if we should (and can) use Pygments to highlight our
   # output.  Reasons not to include a) being told not to by the
@@ -1763,12 +1808,13 @@
                                 get_lexer_for_filename, \
                                 guess_lexer
-    if not encoding:
-      encoding = 'guess'
-      if cfg.options.detect_encoding:
-        try:
-          import chardet
-          encoding = 'chardet'
-        except (SyntaxError, ImportError):
-          pass
 
     # First, see if there's a Pygments lexer associated with MIME_TYPE.
@@ -1807,16 +1853,4 @@
   if not pygments_lexer:
 
-    # If allowed by configuration, try to detect the source encoding
-    # for this file.  We'll assemble a block of data from the file
-    # contents to do so... 1024 bytes should be enough.
-    if not encoding and cfg.options.detect_encoding:
-      block_size = 0
-      text_block = ''
-      for i in range(len(file_lines)):
-        text_block = text_block + file_lines[i]
-        if len(text_block) >= 1024:
-          break
-      encoding = detect_encoding(text_block)
-
     # Built output data comprised of marked-up and possibly-transcoded
     # source text lines wrapped in (possibly dummy) vclib.Annotation

```

**Comment by cmpilato** [comment]
> I had a lot of problems using code which had comments in my native language (encoded in ISO-8859-1),
> because UTF-8 was always chosen as default encoding. I have changed the code a bit and added
> more configuration switches to allow the administrator to have more influence in the encoding that is being used.
> 
> I have also found a spot where the config value detect_encoding was not queries correctly.
> Also, another bug was that the encoding-search-algorithm was only available in the plain text files, not in the pygmentized code.

@danielmarschall Can you flesh out these bug reports with a bit more depth?  For example, the reason that the pygmentized code line doesn't do encoding detection is because Pygments itself will do so on our behalf.  (ViewVC's encoding detection approach was, if I recall correctly, modeled after the one Pygments uses.)

**Comment by futatuki** [comment]
I've read what pygments do for charset/encoding detection by using chardet, and it is nearly same viewvc.py do. It passes top of 1024 bytes (regardless of sequence termination of multi-bytes character) to chardet.detect() and doesn't care 'confidence'.  So, I think that doing charset/encoding detection and transcoding ourself before passing text to pygment is not so bad, and may prefer for Python 3 porting, to reduce bytes object context to care. (and for errors parameter for decode(), I prefer other than 'replace')

But I make an objection for default setting ISO8859-1 as fallback encoding, I think it should be still UTF-8 objectively, though I like EUC-JP :-)

**Comment by danielmarschall** [comment]
The problem with chardet is that the first 1024 bytes are mostly ASCII. For example, in Delphi source, I have hundreds of line interface definitions, and then the actual code. And in the code, I usually write German comments which are ISO8859-1, and these are malformed because chardet only found ASCII in the first 1024 bytes.

The fallback should be defined by the sysadmin. Utf8 would be worthless for my scenario, because all UTF8 files usually have a BOM, so it is unlikely that they fall into the fallback anyway. Everything which don't have a BOM is ISO8859-1 for me. When I write PHP, C, Delphi etc, nothing is UTF8. Everything is ANSI/ISO8859-1. I even think the IDEs and compilers of C and Delphi don't support Unicode source code.

**Comment by futatuki** [comment]
I agree that detecting encoding with only first 1024 bytes is not good, so I think using chardet.UniversalDetector for incremental detection during read file content and make file_lines.

> The fallback should be defined by the sysadmin. 
Yes, so I think we should keep neutral for the default value in the viewvc.conf.dist file.

**Comment by futatuki** [comment]
> I think it should be still UTF-8 objectively, ...

I reconsidered It, then It doesn't seem to be neutral, as sys.getdefaultencoding() on Python 2 returns  'ascii' by default, and default set to UTF-8 is quite unhappy, for incorrect grouping of bytes sequence per character (for both of charset using single byte encoding only and one using multi-bytes encoding).

The reason I said 'UTF-8' for default, is simply sys.getdefaultencoding() on Python 3 returns 'utf-8' by default. I'm very sorry.

Another idea to ease this problem, is give a chance to correct to users, to let them specify a charset/encoding.

**Comment by cmpilato** [comment]
I was responsible for implementing deployment of ViewVC (with CVS and Subversion) and CollabNet, and I can tell you after years and years of dealing with real-world customer data -- there is no other way to deal with CVS/RCS metadata than to let the owner of the data declare the character sets and encoding used.  And even then, it's not a sure thing -- a single repository can have data from different character encodings.  As @futatuki notes, auto-detection will only get you so far because there is much common ground between various encodings and generally not a large sample set to work with.  So the only sane approach is "auto-detect where you can (only settling for 100% detection confidence), allow the repository owner to specify a default encoding, and pray for the best."

**Comment by cmpilato** [comment]
Revisiting this topic today and reflecting on too many years of supporting ViewVC with CVS and Subversion in the corporate worlds (with English, European, and CJK language usage).

Nothing beats the repository policies that @futatuki mentions ("must use Shift-JIS for this repository ... must use iso-2022-jp for that one") when it comes to the anything-goes approach that CVS/RCS take to metadata and file content.  Subversion vastly improves this because the metadata is UTF8 and your can add `; charset=whatever` to the `svn:mime-type` for file content.

So I'm actually quite happy with the code as it sits today, but would accept a new configuration option for specifying the fallback character encoding.  With that in place (which would include the ability to configure per-root configuration overrides of the same), I think we've arguably done the best that we reasonably and performantly can do.

**Comment by futatuki** [comment]
I think this issue was solved in 82e8ffb on master.
Is there anything missing ?

**Comment by futatuki** [comment]
With PR #299,  the default charset/encoding of file contents is switch to the charset/encoding of the locale of which ViewVC is start, and you can overriding it by using config parameter 'default_encoding'

I close this issue with that, however if there is someone who think this issue is not resolved yet, please reopen.
